pkg/forwarder/worker.go

```diff
@@ -60,43 +63,75 @@ func (w *Worker) Stop() {
 func (w *Worker) Start() {
 	go func() {
 		for {
+			// handling high priority transactions first
+			select {
+			case t := <-w.HighPrio:
+				if w.callProcess(t) == nil {
+					continue
+				}
+				return
+			case <-w.stopChan:
+				return
+			default:
+			}
+
 			select {
-			case t := <-w.InputChan:
-				ctx, cancel := context.WithCancel(context.Background())
-
-				done := make(chan interface{})
-				go func() {
-					w.process(ctx, t)
-					done <- nil
-				}()
-
-				select {
-				case <-done:
-					// wait for the Transaction process to be over
-				case <-w.stopChan:
-					// cancel current Transaction if we need to stop the worker
-					cancel()
+			case t := <-w.HighPrio:
+				if w.callProcess(t) != nil {
+					return
+				}
+			case t := <-w.LowPrio:
+				if w.callProcess(t) != nil {
 					return
 				}
-				cancel()
 			case <-w.stopChan:
 				return
 			}
 		}
 	}()
 }
 
+// callProcess will process a transaction and cancel it if we need to stop the
+// worker.
+func (w *Worker) callProcess(t Transaction) error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	done := make(chan interface{})
+	go func() {
+		w.process(ctx, t)
+		done <- nil
+	}()
+
+	select {
+	case <-done:
+		// wait for the Transaction process to be over
+	case <-w.stopChan:
+		// cancel current Transaction if we need to stop the worker
+		cancel()
+		return fmt.Errorf("Worker was requested to stop")
+	}
+	cancel()
+	return nil
+}
+
 func (w *Worker) process(ctx context.Context, t Transaction) {
+	requeue := func() {
+		t.Reschedule()
+		select {
+		case w.RequeueChan <- t:
+		default:
+			log.Errorf("workers retry queue is full: dropping transaction !")
```

Comment:
referring to the `retry queue` here might be confusing if we're debugging this, how about something like `dropping transaction: the transaction retry goroutine is too busy handling other transactions`

Not sure my version is better, let me know what you think